sherpa-onnx/csrc/offline-funasr-nano-model.cc

```diff
@@ -0,0 +1,1037 @@
+// Convert IEEE 754 single-precision (32-bit) float to half-precision (16-bit)
+// float. Handles overflow (clamped to infinity), underflow (clamped to zero),
+// and normal values with proper rounding.
+static inline uint16_t FloatToHalfBits(float f) {
+  uint32_t x;
+  std::memcpy(&x, &f, sizeof(x));
+  const uint32_t sign = (x >> 16) & 0x8000u;
+  const int32_t  exp  = static_cast<int32_t>((x >> 23) & 0xFFu);
+  const uint32_t mant = x & 0x007FFFFFu;
+  if (exp == 255) {
+    if (mant == 0) return static_cast<uint16_t>(sign | 0x7C00u);
+    return static_cast<uint16_t>(sign | 0x7C00u | (mant ? 0x1u : 0));
+  }
+  int32_t new_exp = exp - 127 + 15;
+  if (new_exp >= 31) {
+    return static_cast<uint16_t>(sign | 0x7C00u);
+  } else if (new_exp <= 0) {
+    if (new_exp < -10) {
+      return static_cast<uint16_t>(sign);
+    }
+    uint32_t m = mant | 0x00800000u;
+    int32_t shift = 14 - new_exp;
+    uint32_t half_m = m >> shift;
+    if ((m >> (shift - 1)) & 1u) {
+      half_m += 1;
+    }
+    return static_cast<uint16_t>(sign | (half_m & 0x03FFu));
+  } else {
+    uint16_t half_exp = static_cast<uint16_t>(new_exp << 10);
+    uint32_t half_m   = mant >> 13;
+    if (mant & 0x00001000u) {
+      half_m += 1;
+      if (half_m == 0x0400u) {
+        half_m = 0;
+        half_exp = static_cast<uint16_t>((new_exp + 1) << 10);
+        if ((half_exp >> 10) >= 31) {
+          return static_cast<uint16_t>(sign | 0x7C00u);
+        }
+      }
+    }
+    return static_cast<uint16_t>(sign | half_exp | (half_m & 0x03FFu));
+  }
+}
+
+// Calculate the total number of elements from a tensor shape.
+static inline size_t NumelFromShape(const std::vector<int64_t> &shape) {
+  if (shape.empty()) return 0;
+  size_t n = 1;
+  for (auto d : shape) {
+    if (d <= 0) return 0;
+    n *= static_cast<size_t>(d);
+  }
+  return n;
+}
+
+static inline bool TensorIsCpu(const Ort::Value &v) {
+  if (!v.IsTensor()) return true;
+  auto mi = v.GetTensorMemoryInfo();
+  return mi.GetDeviceType() == OrtMemoryInfoDeviceType_CPU;
+}
+
+static inline void AssertTensorIsCpu(const Ort::Value &v, const char *what) {
+  if (!v.IsTensor()) return;
+  auto mi = v.GetTensorMemoryInfo();
+  if (mi.GetDeviceType() != OrtMemoryInfoDeviceType_CPU) {
+    SHERPA_ONNX_LOGE("%s: expected CPU tensor but got device_type=%d device_id=%d",
+                     what, (int)mi.GetDeviceType(), mi.GetDeviceId());
+    SHERPA_ONNX_EXIT(-1);
+  }
+}
+
+static inline std::string ToLower(std::string s) {
+  std::transform(s.begin(), s.end(), s.begin(),
+                 [](unsigned char c) -> char {
+                   return static_cast<char>(std::tolower(c));
+                 });
+  return s;
+}
+
+static inline bool IsCudaProvider(const std::string &provider) {
+  auto p = ToLower(provider);
+  // Keep it conservative. We only enable IO binding policy below when we
+  // are on CUDA; other EPs keep the existing behavior.
+  return p == "cuda" || (p.size() > 4 && p.rfind("cuda", 0) == 0);
```

Comment:
```suggestion
  return p == "cuda" || (p.size() > 4 && p.find("cuda") == 0);
```

It is more readable to use `find` instead of `rfind()` here.